File: examples/object-detection/multi-camera-people-detection-and-tracking/python/utils/sample_utils.py
"""Sample and payload helpers for the Python multi-camera example."""

from __future__ import annotations

import json
import struct
from typing import Any

from .tracker import TrackedDetection


def filter_person_detections(boxes: list[dict], person_class_id: int = 0) -> list[dict]:
    return [box for box in boxes if int(box.get("class_id", -1)) == int(person_class_id)]


def make_optiview_tracking_json(
    timestamp_ms: int,
    frame_id: str,
    tracked: list[TrackedDetection],
) -> str:
    objects = []
    for det in tracked:
        objects.append(
            {
                "id": f"track_{det.track_id}",
                "track_id": int(det.track_id),
                "label": "person",
                "confidence": float(det.score),
                "bbox": [
                    float(det.x1),
                    float(det.y1),
                    float(max(0.0, det.x2 - det.x1)),
                    float(max(0.0, det.y2 - det.y1)),
                ],
            }
        )
    return json.dumps(
        {
            "type": "object-detection",
            "timestamp": int(timestamp_ms),
            "frame_id": str(frame_id),
            "data": {"objects": objects},
        }
    )


def iter_tensors(pyneat: Any, sample: Any):
    if getattr(sample, "kind", None) == pyneat.SampleKind.Tensor and getattr(sample, "tensor", None) is not None:
        yield sample.tensor
    for field in getattr(sample, "fields", []):
        yield from iter_tensors(pyneat, field)


def first_tensor(pyneat: Any, sample: Any) -> Any:
    if sample is None:
        return None
    if getattr(sample, "kind", None) == pyneat.SampleKind.Tensor and getattr(sample, "tensor", None) is not None:
        return sample.tensor
    for tensor in iter_tensors(pyneat, sample):
        return tensor
    return None


def extract_bbox_payload(pyneat: Any, sample: Any) -> bytes | None:
    stack = [sample]
    while stack:
        current = stack.pop()
        stack.extend(reversed(list(getattr(current, "fields", []))))
        if getattr(current, "kind", None) != pyneat.SampleKind.Tensor:
            continue
        tensor = getattr(current, "tensor", None)
        if tensor is None:
            continue
        fmt = (getattr(current, "payload_tag", "") or getattr(current, "format", "") or "").upper()
        if fmt and fmt != "BBOX":
            continue
        try:
            payload = tensor.copy_payload_bytes()
        except Exception:
            continue
        if payload:
            return payload
    return None


def parse_bbox_payload(payload: bytes | None, img_w: int, img_h: int) -> list[dict]:
    if not payload or len(payload) < 4:
        return []
    count = min(struct.unpack_from("<I", payload, 0)[0], (len(payload) - 4) // 24)
    boxes: list[dict] = []
    offset = 4
    for _ in range(count):
        x, y, w, h, score, class_id = struct.unpack_from("<iiiifi", payload, offset)
        offset += 24
        x1 = max(0.0, min(float(img_w), float(x)))
        y1 = max(0.0, min(float(img_h), float(y)))
        x2 = max(0.0, min(float(img_w), float(x + w)))
        y2 = max(0.0, min(float(img_h), float(y + h)))
        if x2 <= x1 or y2 <= y1:
            continue
        boxes.append(
            {
                "x1": x1,
                "y1": y1,
                "x2": x2,
                "y2": y2,
                "score": float(score),
                "class_id": int(class_id),
            }
        )
    return boxes


def tensor_bgr_from_sample(runtime, sample: Any):
    tensor = first_tensor(runtime.pyneat, sample)
    if tensor is None:
        raise RuntimeError("no tensor payload found in decoded RTSP sample")
    arr = runtime.np.asarray(tensor.to_numpy(copy=True))
    if arr.ndim == 4 and arr.shape[0] == 1:
        arr = arr[0]
    if arr.ndim != 3:
        raise RuntimeError(f"unexpected decoded tensor shape: {arr.shape}")
    if arr.dtype != runtime.np.uint8:
        arr = runtime.np.clip(arr, 0, 255).astype(runtime.np.uint8)
    return runtime.np.ascontiguousarray(arr)

File: examples/tracking/multi-stream-people-tracker/cpp/utils/sample_utils.cpp
#include "examples/tracking/multi-stream-people-tracker/cpp/utils/sample_utils_api.cpp"

#include "neat/runtime.h"
#include "support/object_detection/obj_detection_utils.h"

#include <algorithm>
#include <stdexcept>
#include <vector>

namespace multi_stream_people_tracker {
namespace {

const simaai::neat::Tensor* first_tensor_impl(const simaai::neat::Sample& sample) {
  if (sample.kind == simaai::neat::SampleKind::Tensor && sample.tensor.has_value()) {
    return &sample.tensor.value();
  }
  if (sample.kind == simaai::neat::SampleKind::TensorSet && !sample.tensors.empty()) {
    return &sample.tensors.front();
  }
  for (const auto& field : sample.fields) {
    if (const auto* tensor = first_tensor_impl(field); tensor != nullptr) {
      return tensor;
    }
  }
  return nullptr;
}

bool extract_bbox_payload_impl(const simaai::neat::Sample& sample,
                               std::vector<std::uint8_t>& payload, std::string& error) {
  if (sample.kind == simaai::neat::SampleKind::Bundle) {
    for (const auto& field : sample.fields) {
      if (extract_bbox_payload_impl(field, payload, error)) {
        return true;
      }
    }
    error = "bundle missing BBOX field";
    return false;
  }
  if (sample.kind == simaai::neat::SampleKind::TensorSet && !sample.tensors.empty()) {
    simaai::neat::Sample tensor_sample = sample;
    tensor_sample.kind = simaai::neat::SampleKind::Tensor;
    tensor_sample.tensor = sample.tensors.front();
    tensor_sample.tensors.clear();
    return objdet::extract_bbox_payload(tensor_sample, payload, error);
  }
  return objdet::extract_bbox_payload(sample, payload, error);
}

} // namespace

std::vector<Detection> filter_person_detections(const std::vector<Detection>& boxes,
                                                int person_class_id) {
  std::vector<Detection> filtered;
  for (const auto& box : boxes) {
    if (box.class_id == person_class_id) {
      filtered.push_back(box);
    }
  }
  return filtered;
}

InsightTrackingPayload
make_insight_tracking_detection(const std::vector<TrackedDetection>& tracked) {
  InsightTrackingPayload payload;
  payload.tracks.reserve(tracked.size());
  for (const auto& det : tracked) {
    sima_examples::MetadataBox track;
    track.id = std::to_string(det.track_id);
    track.label = "person";
    track.confidence = det.score;
    track.x = det.x1;
    track.y = det.y1;
    track.w = std::max(0.0f, det.x2 - det.x1);
    track.h = std::max(0.0f, det.y2 - det.y1);
    payload.tracks.push_back(track);
  }
  return payload;
}

std::vector<std::uint8_t> extract_bbox_payload(const simaai::neat::Sample& sample) {
  std::vector<std::uint8_t> payload;
  std::string error;
  if (!extract_bbox_payload_impl(sample, payload, error)) {
    return {};
  }
  return payload;
}

std::vector<Detection> parse_bbox_payload(const std::vector<std::uint8_t>& payload, int img_w,
                                          int img_h) {
  const auto parsed = objdet::parse_boxes_lenient(payload, img_w, img_h, 0);
  std::vector<Detection> out;
  out.reserve(parsed.size());
  for (const auto& box : parsed) {
    out.push_back(Detection{
        box.x1,
        box.y1,
        box.x2,
        box.y2,
        box.score,
        box.class_id,
    });
  }
  return out;
}

cv::Mat tensor_rgb_from_sample(const simaai::neat::Sample& sample) {
  const auto* tensor = first_tensor_impl(sample);
  if (tensor == nullptr) {
    throw std::runtime_error("no tensor payload found in decoded RTSP sample");
  }
  return tensor->to_cv_mat_copy(simaai::neat::ImageSpec::PixelFormat::RGB);
}

} // namespace multi_stream_people_tracker
